winml/lib/Api.Core/WinMLAdapter.cpp

```diff
@@ -32,6 +31,38 @@
 using namespace winrt::Windows::AI::MachineLearning;
 
 namespace Windows::AI::MachineLearning::Adapter {
+// Define winml trace logging provider with WinML GUID
+#ifdef LAYERING_DONE
+TRACELOGGING_DEFINE_PROVIDER(
+    winml_trace_logging_provider,
+    WINML_PROVIDER_DESC,
+    WINML_PROVIDER_GUID,
+    TraceLoggingOptionMicrosoftTelemetry());
+#else
+TRACELOGGING_DEFINE_PROVIDER(
+    winml_trace_logging_provider,
+    WINML_PROVIDER_DESC,
+    WINML_PROVIDER_GUID);
+#endif
+
+// Class to unregister trace logging provider at shutdown.
+// Only one static instance to be created for the lifetime of the program.
+class WinMLTraceLoggingProviderManager {
+ public:
+  static WinMLTraceLoggingProviderManager& Register() {
+    const HRESULT etw_status = TraceLoggingRegister(winml_trace_logging_provider);
+    if (FAILED(etw_status)) {
+      throw std::runtime_error("WinML TraceLogging registration failed. Logging will be broken: " + std::to_string(etw_status));
+    }
+    // return an instance that is just used to unregister as the program exits
```

Comment:
I think the comment is misleading, you don't need an instance to unregister (the static variable will always have its destructor called at program exit, and only at program exit, right?). You could as well return `void` and the result would be the same (correct me if I'm wrong)

Reply:
Agreed! I modified it a little bit so that the registration is in the constructor.